networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/gossip/subnets/SubnetScorer.java

```diff
@@ -45,43 +56,93 @@ public int scoreExistingPeer(final NodeId peerId) {
         attSubscriptions,
         syncCommitteeSubscriptions,
         dataColumnSidecarSubscriptions,
-        this::scoreSubnetForExistingPeer);
+        this::scoreSubnetForExistingPeer,
+        new SelectedCandidateSubnetCountChanges());
   }
 
   @Override
-  public int scoreCandidatePeer(final DiscoveryPeer candidate) {
+  public List<DiscoveryPeer> selectCandidatePeers(
+      final List<DiscoveryPeer> candidates, final int maxToSelect) {
+    final SelectedCandidateSubnetCountChanges subnetChanges =
+        new SelectedCandidateSubnetCountChanges();
+
+    List<DiscoveryPeer> selectedPeers = new ArrayList<>();
+    List<DiscoveryPeer> remainingPeers = new ArrayList<>(candidates);
+    while (selectedPeers.size() < maxToSelect) {
+      Optional<DiscoveryPeer> maybeHighestScoringPeer =
+          remainingPeers.stream()
+              .max(Comparator.comparing(peer -> scoreCandidatePeer(peer, subnetChanges)));
+      if (maybeHighestScoringPeer.isEmpty()) {
+        break;
+      }
+      DiscoveryPeer highestScoringPeer = maybeHighestScoringPeer.get();
+      selectedPeers.add(highestScoringPeer);
+      remainingPeers.remove(highestScoringPeer);
+
+      peerSubnetSubscriptions
```

Comment:
`peerSubnetSubscriptions` have data only for existing peers. It doesn't contain any data for candidate peers. LibP2P `NodeId` and discovery `NodeId` are different concepts with similar name, that's why all methods in `peerSubnetSubscriptions` are tied to LibP2P's `NodeId`. We could use only data that were used before:
```java
candidate.getPersistentAttestationSubnets(),
candidate.getSyncCommitteeSubnets(),
peerSubnetSubscriptions.getDataColumnSidecarSubnetSubscriptionsByNodeId(
UInt256.fromBytes(candidate.getNodeId()
```